packages/store/src/decorators/state.ts

```diff
@@ -1,42 +1,37 @@
-import { ensureStoreMetadata } from '../internal/internals';
-import { StoreOptions, META_KEY } from '../symbols';
-
-const stateNameRegex = new RegExp('^[a-zA-Z0-9_]+$');
-
-/**
- * Error message
- * @ignore
- */
-export const stateNameErrorMessage = (name: string) =>
-  `${name} is not a valid state name. It needs to be a valid object property name.`;
+import {
+  ensureStoreMetadata,
+  MetaDataModel,
+  MutateMetaOptions,
+  StateClass
+} from '../internal/internals';
+import { META_KEY, META_OPTIONS_KEY, StoreOptions } from '../symbols';
+import { StoreValidators } from '../utils/store-validators';
 
 /**
  * Decorates a class with ngxs state information.
  */
 export function State<T>(options: StoreOptions<T>) {
-  return function(target: any) {
-    const meta = ensureStoreMetadata(target);
-
-    // Handle inheritance
-    if (Object.getPrototypeOf(target).hasOwnProperty(META_KEY)) {
-      const parentMeta = Object.getPrototypeOf(target)[META_KEY];
-
-      meta.actions = {
-        ...meta.actions,
-        ...parentMeta.actions
-      };
+  function mutateMetaData(params: MutateMetaOptions<T>): void {
+    const { meta, stateClass, optionsWithInheritance } = params;
+    const { children, defaults, name } = optionsWithInheritance;
+    StoreValidators.checkCorrectStateName(name);
+
+    if (stateClass.hasOwnProperty(META_KEY)) {
+      const parentMeta: Partial<MetaDataModel> = stateClass[META_KEY] || {};
+      meta.actions = { ...meta.actions, ...parentMeta.actions };
     }
 
-    meta.children = options.children;
-    meta.defaults = options.defaults;
-    meta.name = options.name;
-
-    if (!options.name) {
-      throw new Error(`States must register a 'name' property`);
-    }
-
-    if (!stateNameRegex.test(options.name)) {
-      throw new Error(stateNameErrorMessage(options.name));
-    }
+    meta.children = children;
+    meta.defaults = defaults;
+    meta.name = name;
+  }
+
+  return (target: StateClass): void => {
+    const meta: MetaDataModel = ensureStoreMetadata(target);
+    const stateClass: StateClass = Object.getPrototypeOf(target);
```

Comment:
It is the prototype of the target class, therefore is it not a class we have inherited from (or extended)? I was trying to avoid the word `parent` because that could refer to the state graph.
`inherited` is the best I can come up with to describe the extends relationship.

Reply:
done